language/src/regent/optimize_index_launches.t

```diff
@@ -1177,10 +1199,147 @@ local function optimize_loop_body(cx, node, report_pass, report_fail)
+  local bitmask_i = util.mk_expr_index_access(util.mk_expr_id_rawref(bitmask), util.mk_expr_id(idx), std.rawref(&bool))
+  stats:insert(util.mk_stat_assignment(bitmask_i, util.mk_expr_constant(false, bool)))
+
+  local values = terralib.newlist()
+  values:insert(util.mk_expr_constant(0, int32))
+  values:insert(util.mk_expr_id(volume))
+
+  return util.mk_stat_for_num(idx, values, util.mk_block(stats))
+end
+
+-- Computes: pf.z + pf.y * (bounds.hi.z - bounds.lo.z + 1) + pf.x * (bounds.hi.z - bounds.lo.z + 1) * (bounds.hi.y - bounds.lo.y + 1)
+local function collapse_projection_functor(pf, dim, bounds)
+  local index_types = { [0] = std.int1d, std.int1d, std.int2d, std.int3d }
+
+  if dim == 0 or dim == 1 then
+    return pf
+  else
+    local x = util.mk_expr_field_access(pf, "x", int32)
+    local y = util.mk_expr_field_access(pf, "y", int32)
+  
+    local y_extent = util.mk_expr_binary("+", util.mk_expr_binary("-", util.mk_expr_field_access(util.mk_expr_field_access(bounds, "hi", index_types[dim]), "y", int32), util.mk_expr_field_access(util.mk_expr_field_access(bounds, "lo", index_types[dim]), "y", int32)), util.mk_expr_constant(1, int32))
+  
+    if dim == 2 then
+      return util.mk_expr_binary("+", util.mk_expr_binary("*", x, y_extent), y)
+
+    elseif dim == 3 then
+      local z = util.mk_expr_field_access(pf, "z", int32)
+      local z_extent = util.mk_expr_binary("+", util.mk_expr_binary("-", util.mk_expr_field_access(util.mk_expr_field_access(bounds, "hi", index_types[dim]), "z", int32), util.mk_expr_field_access(util.mk_expr_field_access(bounds, "lo", index_types[dim]), "z", int32)), util.mk_expr_constant(1, int32))
+
+      return util.mk_expr_binary("+", util.mk_expr_binary("*", x, util.mk_expr_binary("*", y_extent, z_extent)), util.mk_expr_binary("+", util.mk_expr_binary("*", y, z_extent), z))
+
+    else
+      assert(false)
+    end
+  end
+end
+
+local function mk_duplicates_check(preamble, value, index_expr, bitmask, conflict, verdict, loop_var, bounds, dim, volume)
+  local stats = terralib.newlist()
+
+  -- Compute value = index_expr(i)
+  preamble:map(function(stat) stats:insert(stat) end)
+  stats:insert(util.mk_stat_assignment(util.mk_expr_id_rawref(value), collapse_projection_functor(index_expr, dim, bounds)))
+
+  local then_block = terralib.newlist()
+
+  local bitmask_value = util.mk_expr_index_access(util.mk_expr_id(bitmask), util.mk_expr_id(value), std.rawref(&bool))
+  then_block:insert(util.mk_stat_assignment(util.mk_expr_id_rawref(conflict), bitmask_value))
+
+  then_block:insert(util.mk_stat_assignment(bitmask_value, util.mk_expr_constant(true, bool)) )
+
+  local set_verdict = util.mk_stat_assignment(util.mk_expr_id_rawref(verdict), util.mk_expr_constant(true, bool))
+  then_block:insert(util.mk_stat_if(util.mk_expr_id(conflict), terralib.newlist { set_verdict, util.mk_stat_break() }))
+
+  -- Bounds check
+  local cond = util.mk_expr_binary("<", util.mk_expr_id(value), util.mk_expr_id(volume))
```

Comment:
Should we also do a check that it's `>= 0`?

Reply:
Pushed.